// eslint-plugin-lodash/lib/rules/no-unbound-this.js
/**
 * @fileoverview Rule to check that all uses of `this` inside collection methods are bound
 */
'use strict';
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    create: function create(context) {
        var _require = require('../util/lodashUtil');

        var getLodashMethodCallExpVisitor = _require.getLodashMethodCallExpVisitor;
        var getLodashContext = _require.getLodashContext;

        var _require2 = require('../util/methodDataUtil');

        var getCollectionMethods = _require2.getCollectionMethods;

        var _require3 = require('../util/astUtil');

        var isFunctionExpression = _require3.isFunctionExpression;

        var assign = require('lodash/assign');
        var collectionMethods = void 0;
        var funcInfos = new Map();
        var currFuncInfo = {
            thisUses: []
        };
        var lodashContext = getLodashContext(context);
        return assign({
            'CallExpression:exit': getLodashMethodCallExpVisitor(lodashContext, function (node, iteratee, _ref) {
                var method = _ref.method;
                var version = _ref.version;

                collectionMethods = collectionMethods || new Set(getCollectionMethods(version).concat(['forEach', 'forEachRight']));
                if (collectionMethods.has(method) && funcInfos.has(iteratee)) {
                    var _funcInfos$get = funcInfos.get(iteratee);

                    var thisUses = _funcInfos$get.thisUses;

                    if (isFunctionExpression(iteratee) && thisUses.length) {
                        thisUses.forEach(function (thisNode) {
                            context.report(thisNode, 'Do not use `this` without binding in collection methods');
                        });
                    }
                }
            }),
            ThisExpression: function ThisExpression(node) {
                currFuncInfo.thisUses.push(node);
            },
            onCodePathStart: function onCodePathStart(codePath, node) {
                currFuncInfo = {
                    upper: currFuncInfo,
                    codePath: codePath,
                    thisUses: []
                };
                funcInfos.set(node, currFuncInfo);
            },
            onCodePathEnd: function onCodePathEnd() {
                currFuncInfo = currFuncInfo.upper;
            }
        }, lodashContext.getImportVisitors());
    }
};
