// eslint-plugin-lodash/lib/rules/prefer-lodash-method.js
/**
 * @fileoverview Rule to check if there's a method in the chain start that can be in the chain
 */
'use strict';

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
    meta: {
        schema: [{
            type: 'object',
            properties: {
                ignoredMethods: {
                    type: 'array',
                    items: {
                        type: 'string'
                    }
                },
                ignoredObjects: {
                    type: 'array',
                    items: {
                        type: 'string'
                    }
                }
            }
        }]
    },

    create: function create(context) {
        var _require = require('../util/lodashUtil');

        var getLodashContext = _require.getLodashContext;
        var isNativeCollectionMethodCall = _require.isNativeCollectionMethodCall;
        var getLodashMethodCallExpVisitor = _require.getLodashMethodCallExpVisitor;

        var _require2 = require('../util/astUtil');

        var getMethodName = _require2.getMethodName;
        var getCaller = _require2.getCaller;

        var get = require('lodash/get');
        var includes = require('lodash/includes');
        var matches = require('lodash/matches');
        var some = require('lodash/some');
        var map = require('lodash/map');
        var assign = require('lodash/assign');
        var ignoredMethods = get(context, ['options', 0, 'ignoreMethods'], []);
        var ignoredObjects = get(context, ['options', 0, 'ignoreObjects'], []);
        var usingLodash = new Set();

        var lodashContext = getLodashContext(context);

        function isNonNullObjectCreate(callerName, methodName, arg) {
            return callerName === 'Object' && methodName === 'create' && get(arg, 'value') !== null;
        }

        function isStaticNativeMethodCall(node) {
            var staticMethods = {
                Object: ['assign', 'keys', 'values'],
                Array: ['isArray']
            };
            var callerName = get(node, 'callee.object.name');
            var methodName = getMethodName(node);
            return callerName in staticMethods && includes(staticMethods[callerName], methodName) || isNonNullObjectCreate(callerName, methodName, node.arguments[0]);
        }

        function isNativeStringMethodCall(node) {
            return includes(['endsWith', 'includes', 'padEnd', 'padStart', 'startsWith'], getMethodName(node));
        }

        function canUseLodash(node) {
            return isNativeCollectionMethodCall(node) || isStaticNativeMethodCall(node) || isNativeStringMethodCall(node);
        }

        function getTextOfNode(node) {
            if (node) {
                if (node.type === 'Identifier') {
                    return node.name;
                }
                return context.getSourceCode().getText(node);
            }
        }

        function someMatch(patterns, str) {
            return str && some(patterns, function (pattern) {
                return str.match(pattern);
            });
        }

        function shouldIgnore(node) {
            return someMatch(ignoredMethods, getMethodName(node)) || someMatch(ignoredObjects, getTextOfNode(getCaller(node)));
        }
        return assign({
            CallExpression: getLodashMethodCallExpVisitor(lodashContext, function (node) {
                usingLodash.add(node);
            }),
            'CallExpression:exit': function CallExpressionExit(node) {
                if (!usingLodash.has(node) && !shouldIgnore(node) && canUseLodash(node)) {
                    context.report(node, 'Prefer \'_.' + getMethodName(node) + '\' over the native function.');
                }
            }
        }, lodashContext.getImportVisitors());
    }
};
